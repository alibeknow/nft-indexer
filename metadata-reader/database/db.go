package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
