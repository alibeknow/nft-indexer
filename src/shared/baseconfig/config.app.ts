import { IsEnum, IsOptional, IsString } from 'class-validator';

export enum LogLevel {
  DEBUG = 'debug',
  INFO = 'info',
}

export interface IAppConfig {
  version?: string;
  logLevel: string;
}

export class AppConfigValidator implements IAppConfig {
  @IsString()
  @IsOptional()
  readonly version?: string;

  @IsEnum(LogLevel)
  readonly logLevel!: LogLevel;
}

export const getAppConfig = (): IAppConfig => ({
  version: process.env.APP_VERSION,
  logLevel: process.env.LOG_LEVEL || LogLevel.INFO,
});

