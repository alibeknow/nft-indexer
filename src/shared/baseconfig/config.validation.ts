import { plainToClass } from 'class-transformer';
import { validateSync } from 'class-validator';
import { BaseConfigValidator } from './config.base';

export function validate<K, T extends BaseConfigValidator>(config: K, ValidationClass: new () => T): T {
  const validatedConfig = plainToClass(
    ValidationClass,
    config,
    { enableImplicitConversion: true },
  );
  const errors = validateSync(validatedConfig, { skipMissingProperties: false });

  if (errors.length > 0) {
    throw new Error(errors.toString());
  }

  return validatedConfig;
}
