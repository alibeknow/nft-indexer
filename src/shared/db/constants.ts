export const MONGO_TOKENS_COLLECTION_NAME = 'tokens';
export const MONGO_CONTRACTS_COLLECTION_NAME = 'contracts';
export const MONGO_METADATA_COLLECTION_NAME = 'metadata';
export const MONGO_EVENTS_COLLECTION_NAME = 'events';
export const MONGO_EVENTS_READER_STATE_LOG_COLLECTION_NAME = 'events_reader_state_log';
export const MONGO_BLOCKS_QUEUE_COLLECTION_NAME = 'blocks_queue';
export const MONGO_LAZY_MINT_BLOCKS_QUEUE_COLLECTION_NAME = 'lazy_mint_blocks_queue';
export const MONGO_SERVER_DUPLICATE_KEY_ERROR_CODE = 11000;
