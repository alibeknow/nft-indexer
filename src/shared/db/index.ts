export * from './constants';
export * from './db.class';
export * from './db.deprecated';
export * from './db.config';
export * from './db.module';
export * from './db.provider';
export * from './mongo-sort';

