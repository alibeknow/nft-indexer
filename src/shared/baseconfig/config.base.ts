import { ExplorersClientConfigValidator, IExplorersClientConfig, getExplorersClientConfig } from '@shared/explorers-client';
import { ValidateNested } from 'class-validator';
import { Type } from 'class-transformer';

import { AlchemyConfigValidator, IAlchemyConfig, getAlchemyConfig } from '@shared/alchemy';
import { IWeb3Config, Web3ConfigValidator, getWeb3Config } from '@shared/web3';
import { DatabaseConfigValidator, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { AWSConfigValidator, IAWSConfig, getAWSConfig } from '@shared/aws';
import {
  IMicroservicesConfig, IRabbitConfig, ISqsConfig,
  MicroservicesConfigValidator, RabbitConfigValidator, SqsConfigValidator,
  getMicroservicesConfig, getRabbitConfig, getSqsConfig,
} from '@shared/microservices';
import { IOpensearchConfig, OpensearchConfigValidator, getOpensearchConfig } from '@shared/opensearch';
import { AppConfigValidator, IAppConfig, getAppConfig } from './config.app';
import { IOpenseaConfig, OpenseaConfigValidator, getOpenseaConfig } from '@shared/opensea';

export interface IBaseConfig {
  app: IAppConfig;
  alchemy: IAlchemyConfig;
  explorersClient: IExplorersClientConfig;
  aws: IAWSConfig;
  database: IDatabaseConfig;
  microservices: IMicroservicesConfig;
  opensearch: IOpensearchConfig;
  opensea: IOpenseaConfig;
  rabbit: IRabbitConfig;
  sqs: ISqsConfig;
  web3: IWeb3Config;
}

export class BaseConfigValidator implements IBaseConfig {
  @ValidateNested()
  @Type(() => AppConfigValidator)
  readonly app!: AppConfigValidator;

  @ValidateNested()
  @Type(() => AlchemyConfigValidator)
  readonly alchemy!: AlchemyConfigValidator;

  @ValidateNested()
  @Type(() => ExplorersClientConfigValidator)
    explorersClient!: ExplorersClientConfigValidator;

  @ValidateNested()
  @Type(() => AWSConfigValidator)
  readonly aws!: AWSConfigValidator;

  @ValidateNested()
  @Type(() => DatabaseConfigValidator)
  readonly database!: DatabaseConfigValidator;

  @ValidateNested()
  @Type(() => MicroservicesConfigValidator)
  readonly microservices!: MicroservicesConfigValidator;

  @ValidateNested()
  @Type(() => RabbitConfigValidator)
  readonly rabbit!: RabbitConfigValidator;

  @ValidateNested()
  @Type(() => SqsConfigValidator)
  readonly sqs!: SqsConfigValidator;

  @ValidateNested()
  @Type(() => Web3ConfigValidator)
  readonly web3!: Web3ConfigValidator;

  @ValidateNested()
  @Type(() => OpensearchConfigValidator)
  readonly opensearch!: OpensearchConfigValidator;

  @ValidateNested()
  @Type(() => OpenseaConfigValidator)
  readonly opensea!: OpenseaConfigValidator;
}

export const getBaseConfig = (): IBaseConfig => ({
  app: getAppConfig(),
  alchemy: getAlchemyConfig(),
  explorersClient: getExplorersClientConfig(),
  aws: getAWSConfig(),
  database: getDatabaseConfig(),
  microservices: getMicroservicesConfig(),
  rabbit: getRabbitConfig(),
  sqs: getSqsConfig(),
  web3: getWeb3Config(),
  opensearch: getOpensearchConfig(),
  opensea: getOpenseaConfig(),
});
