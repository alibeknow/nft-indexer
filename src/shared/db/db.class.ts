import { LazyMintBlockQueue } from '../lazy-mint-blocks-queue';
import { Collection, MongoClient } from 'mongodb';
import { BlockQueue } from '../blocks-queue';
import {
  MONGO_BLOCKS_QUEUE_COLLECTION_NAME,
  MONGO_CONTRACTS_COLLECTION_NAME,
  MONGO_EVENTS_COLLECTION_NAME,
  MONGO_EVENTS_READER_STATE_LOG_COLLECTION_NAME,
  MONGO_LAZY_MINT_BLOCKS_QUEUE_COLLECTION_NAME,
  MONGO_METADATA_COLLECTION_NAME,
  MONGO_TOKENS_COLLECTION_NAME,
} from './constants';
import { Token } from '../tokens';
import { Metadata } from '../metadata';
import { Event, EventsReaderStateLog } from '../events';
import { MongoSort } from './mongo-sort';
import { Contract } from '../contracts';
import { IDatabaseConfig, getMongoDBUri } from './db.config';

export class DBClass {
  private client: MongoClient;
  private directConnection: boolean;
  private sslEnable: boolean;
  private uri: string;
  private dbName: string;

  constructor(databaseConfig: IDatabaseConfig) {
    this.directConnection = databaseConfig.directConection;
    this.sslEnable = databaseConfig.enableSsl;
    this.dbName = databaseConfig.db;
    this.uri = getMongoDBUri(databaseConfig);

    let options = {};

    if(this.directConnection) {
      options = {
        ...options,
        directConnection: true,
      };
    }

    if(this.sslEnable) {
      options = {
        ...options,
        ssl: true,
        sslValidate: false,
        sslCA: `${__dirname  }/../../../cert/rds-combined-ca-bundle.pem`,
      };
    }

    this.client = new MongoClient(this.uri, options);
  }

  public async open(): Promise<MongoClient> {
    return this.client.connect();
  }

  public async close(force: boolean): Promise<void> {
    return this.client.close(force);
  }

  public tokens(): Collection<Token> {
    return this.client.db(this.dbName).collection<Token>(MONGO_TOKENS_COLLECTION_NAME);
  }

  public async createTokensIndexes(): Promise<string> {
    return this.tokens().createIndex({ createdAt: MongoSort.ASC });
  }

  public async createBlockQueueIndexes(): Promise<string> {
    return this.blocksQueue().createIndex({ to: MongoSort.DESC });
  }

  public async createLazyMintBlockQueueIndexes(): Promise<string> {
    return this.lazyMintBlocksQueue().createIndex({ to: MongoSort.DESC });
  }

  public metadata(): Collection<Metadata> {
    return this.client.db(this.dbName).collection<Metadata>(MONGO_METADATA_COLLECTION_NAME);
  }

  public events(): Collection<Event> {
    return this.client.db(this.dbName).collection<Event>(MONGO_EVENTS_COLLECTION_NAME);
  }

  public async eventsIndexes(): Promise<string> {
    return this.events().createIndex({ blockNumber: MongoSort.ASC, logIndex: MongoSort.ASC });
  }

  public eventsReaderStateLog(): Collection<EventsReaderStateLog> {
    return this.client.db(this.dbName).collection<EventsReaderStateLog>(MONGO_EVENTS_READER_STATE_LOG_COLLECTION_NAME);
  }

  public async eventsReaderStateLogIndexes(): Promise<string> {
    return this.eventsReaderStateLog().createIndex({ startAt: MongoSort.DESC });
  }

  public blocksQueue(): Collection<BlockQueue> {
    return this.client.db(this.dbName).collection<BlockQueue>(MONGO_BLOCKS_QUEUE_COLLECTION_NAME);
  }

  public lazyMintBlocksQueue(): Collection<LazyMintBlockQueue> {
    return this.client.db(this.dbName).collection<LazyMintBlockQueue>(MONGO_LAZY_MINT_BLOCKS_QUEUE_COLLECTION_NAME);
  }

  public contracts(): Collection<Contract> {
    return this.client.db(this.dbName).collection<Contract>(MONGO_CONTRACTS_COLLECTION_NAME);
  }
}
