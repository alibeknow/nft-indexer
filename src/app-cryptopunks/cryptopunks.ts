import { ConfigService } from '@nestjs/config';
import dotenv from 'dotenv';
import * as ethers from 'ethers';
import tokenAbi from './token-abi.json';
import dataAbi from './data-abi.json';
import { DBClass, IDatabaseConfig } from '@shared/db';
import { closeDb } from '@shared/close';
import { logger } from '@shared/logger';
import { TokenStandard } from '@shared/tokens/token-standard';
import { Web3Provider } from '@shared/web3';
import { IBaseConfig } from '@shared/baseconfig';
import { Blockchain } from '@shared/blockchain';
import { TokenAttributes, TokensRepo } from '@shared/tokens/repo';
import { ContractsRepo } from '@shared/contracts';
import { DATA_CONTRACT_ADDRESS, TOKEN_CONTRACT_ADDRESS } from './constants';

dotenv.config();

const tokenContractAbi = JSON.stringify(tokenAbi);
const dataContractAbi = JSON.stringify(dataAbi);

export async function processCryptoPunks(baseConfig: IBaseConfig) {
  const dbConfig: IDatabaseConfig = baseConfig.database;
  const db = new DBClass(dbConfig);

  const config = new ConfigService<IBaseConfig>(baseConfig);
  const web3Provider = new Web3Provider<IBaseConfig>(config);
  const provider = web3Provider.provider;

  const cryptoPunksTokens = new ethers.Contract(TOKEN_CONTRACT_ADDRESS, tokenContractAbi, provider);
  const cryptoPunksData = new ethers.Contract(DATA_CONTRACT_ADDRESS, dataContractAbi, provider);

  const contractRepo = new ContractsRepo(db);
  const tokenRepo = new TokensRepo(db);

  const totalSupply = await cryptoPunksTokens.totalSupply();

  const contract = {
    blockchain: Blockchain.ETH,
    address: TOKEN_CONTRACT_ADDRESS,
    type: TokenStandard.CRYPTOPUNKS,
    name: 'CryptoPunks',
    block: 0,
    createdAt: new Date(),
  };

  await db.open();
  logger.info({ msg: 'DB connection is opened' });

  try {
    await contractRepo.insertIfNone(contract);
  } catch (err) {
    logger.error({ msg: 'Error storing contract into DB', error: err });
  }

  let data: TokenAttributes;
  let id = '';
  let metadata = '';

  logger.info({ msg: 'Started processing CryptoPunks data ðŸš€' });

  for (let i = 0; i < totalSupply; i++) {
    id = `eth:${TOKEN_CONTRACT_ADDRESS}:${i}`;

    try {
      metadata = await cryptoPunksData.punkImageSvg(i);
    } catch (err) {
      logger.error({ msg: 'Error getting punk image svg', error: err });
    }

    data = {
      blockchain: Blockchain.ETH,
      contractAddress: TOKEN_CONTRACT_ADDRESS,
      tokenId: id,
      block: 0,
      from: '',
      tokenUri: metadata,
      count: 1,
      createdAt: new Date(),
    };

    try {
      await tokenRepo.save(data);
    } catch (err) {
      logger.error({ msg: 'Error storing token into DB', error: err });
    }
  }

  await closeDb(db);
  logger.info({ msg: 'Processing is finished' });
  logger.info({ msg: 'DB connection is closed' });
}
