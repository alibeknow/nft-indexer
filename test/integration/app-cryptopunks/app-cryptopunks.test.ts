import * as ethers from 'ethers';
import { DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { processCryptoPunks }  from '@app-cryptopunks/cryptopunks';
import { IBaseConfig } from '@shared/baseconfig';
import { fixturePunkImageSvg } from './fixtures/fixture.punkImageSvg';

jest.setTimeout(100000);

jest.spyOn(ethers, 'Contract').mockImplementation(
  jest.fn().mockImplementation(() => {
    return {
      totalSupply: jest
        .fn()
        .mockReturnValue(3),
      punkImageSvg: jest
        .fn()
        .mockReturnValue(fixturePunkImageSvg),
    };
  }),
);

describe('app-cryptopunks', () => {
  let db: DBClass;
  let dbConfig: IDatabaseConfig;

  beforeAll(async () => {
    const baseDbConfig: IDatabaseConfig = getDatabaseConfig();
    dbConfig = { ...baseDbConfig, db: `app-cryptopunks-test-${baseDbConfig.db}` };
    db = new DBClass(dbConfig);
    await db.open();

    await Promise.all([
      db.tokens().deleteMany({}),
      db.contracts().deleteMany({}),
    ]);
  });

  it('should write cryptopunks contract data and tokens to contracts collection', async () => {
    const baseConfig = {
      database: dbConfig,
      web3: {
        protocol: process.env.NODE_PROTOCOL as string,
        host: process.env.NODE_HOST as string,
        username: process.env.NODE_USERNAME,
        password: process.env.NODE_PASSWORD,
      },
    };
    await processCryptoPunks(baseConfig as IBaseConfig);
    const contractsResult = await db.contracts().find({}).project({ createdAt: 0, updatedAt: 0 }).toArray();
    const tokensResult = await db.tokens().find({}).project({ createdAt: 0, updatedAt: 0 }).toArray();

    expect(contractsResult).toMatchSnapshot();
    expect(tokensResult).toMatchSnapshot();
  });

  afterAll(async () => {
    await db.tokens().deleteMany({});
    await db.contracts().deleteMany({});
    await db.close(true);
  });
});

