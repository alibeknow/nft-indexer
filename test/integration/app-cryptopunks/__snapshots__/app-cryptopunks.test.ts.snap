// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-cryptopunks should write cryptopunks contract data and tokens to contracts collection 1`] = `
[
  {
    "_id": "eth:0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB",
    "address": "0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB",
    "block": 0,
    "collectionImage": null,
    "name": "CryptoPunks",
    "type": "CRYPTOPUNKS",
  },
]
`;

exports[`app-cryptopunks should write cryptopunks contract data and tokens to contracts collection 2`] = `
[
  {
    "_id": "eth:0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB:eth:0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB:0",
    "block": 0,
    "count": 1,
    "from": "",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" version="1.2" viewBox="0 0 24 24"><rect x="9" y="5" width="1" height="1"",
  },
  {
    "_id": "eth:0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB:eth:0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB:1",
    "block": 0,
    "count": 1,
    "from": "",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" version="1.2" viewBox="0 0 24 24"><rect x="9" y="5" width="1" height="1"",
  },
  {
    "_id": "eth:0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB:eth:0xb47e3cd837dDF8e4c57F05d70Ab865de6e193BBB:2",
    "block": 0,
    "count": 1,
    "from": "",
    "retryCount": 0,
    "tokenMetadataStatus": "undefined",
    "tokenUri": "data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" version="1.2" viewBox="0 0 24 24"><rect x="9" y="5" width="1" height="1"",
  },
]
`;
